### ISSUE TITLE
Dependency of the number of walkers

### ISSUE THREAD

**Issue by afernandezody** [QMCPACK/qmcpack#2191]
Hello QMCPACK Team,
I've run the same problem (input file is attached) on different CPU types with the same amount of memory (16 GB per core). The funny thing is that the Intel processors (Skylake and Scalable) compute 14 walkers per MPI rank whereas AMD (EPYC2-Rome) is computing only 8 walkers per rank, which was unanticipated. The output files show that the maximum and minimum number of walkers per MPI rank follow the described pattern (maximum of 33 for Intel and minimum of 2 for AMD). A separate attempt on an older Hasswell architecture has shown inconclusive results with the number of walkers fluctuating between 8 and 16 per MPI rank. 
The bottom line is that I was expecting the number of walkers to be consistent for the same problem with MPI ranks and memory. I'm likely missing some element but cannot put my finger on it. Any help or insight would be appreciated.
Thanks.

[inputfile.txt](https://github.com/QMCPACK/qmcpack/files/4030931/inputfile.txt)

**Comment by ye-luo** [comment]
Please provide the input xml, job script and output.

**Comment by afernandezody** [comment]
Hi @ye-luo,
I was uploading the input file as you were posting (had initially forgotten as usually). The command line is simply:
`mpirun -np 8 $HOME/qmcpack-3.8.0/build/bin/qmcpack NiO-example.in.xml > out_Rome8`  [Here NiO-example.in.xml is the input file previously uploaded] (PBS scripts for large clusters follow a similar pattern).
The attached files are the outputs from an Intel architecture and from an AMD architecture. Thanks.


[out_Intel8.txt](https://github.com/QMCPACK/qmcpack/files/4030964/out_Intel8.txt)
[out_Rome8.txt](https://github.com/QMCPACK/qmcpack/files/4030965/out_Rome8.txt)

**Comment by prckent** [comment]
The VMC implementation in QMCPACK does not allow any threads to be idle and requires an equal number of walkers per thread. This ensures perfect load balancing in VMC. Thus an input file specifying walkers=14 (this is the number of walkers per MPI task to use), when run with 8 threads will be rounded down to 8 walkers, i.e. one per thread. This constraint largely exists to protect users from poorly using the underlying hardware. Thus in your input you need to put walkers=8 or walkers=16, or some higher multiple. On an imaginary future processor with 137 cores you could put walkers=137 and run with 137 threads. If you put walkers=13 on this processor the code would presently round up the walker count to 137.

You can see this in your sample outputs as follows. In out_Intel8.txt the start of the VMC block prints "adding 14 walkers...", and "walkers/mpi=14".  You are running with 8 mpi tasks and 2 threads per task. Since 14 is a multiple of 2, the code runs the requested number of walkers.
In out_Rome8.txt you have 8 mpi tasks and 8 threads per task. Your input still requested 14 walkers/mpi but the code rounds this down to 8. The start of the VMC therefore block prints "Adding 8 walkers" etc. Therefore these runs are different and you observe the different behavior that you report.

DMC necessarily allows for a fully variable population count and therefore load imbalance between mpi tasks, threads etc. This also leads to additional variation in the performance. If you are focusing on measuring or improving performance, focusing initially only on the VMC performance is a good idea.

Hopefully the above is clear! Let us know if you have more questions.

**Comment by afernandezody** [comment]
Hi @prckent,
Thank you for your feedback. I understand what you say (the dependency of the number of walkers on the threads and other considerations) but I cannot figure out how the number of threads is set. As my systems don't enable OpenMP, I'm assuming that the software runs on a pure MPI mode and that the number of threads is chosen based on the underlying hardware. Is my assumption correct or is there a way to tell the code how many threads to use? Also, would memory allocation (e.g. NUMA vs. Non-NUMA) have any impact on this parameter?

**Comment by prckent** [comment]
The outputs you sent indicate that QMCPACK was built with threading enabled. See the fifteenth line of output or so "OMP 1st level threads     = 8". Use export OMP_NUM_THREADS=1 or whatever you need before running.

**Comment by afernandezody** [comment]
Hi @prckent,
My recollection is that I needed to enable threading in order to compile (will have to double check later with my notes) even if each MPI rank is single-threaded. The missing piece was that one still has to tell the system about the number of threads or the system will choose it for you (apparently Intel prefers 2 and AMD 8 but there is probably some technical explanation for these choices). Everything is logical once you think about it. I'm running a couple of tests (don't anticipate further issues) and will close the ticket later today. 
Thank you for the help.

**Comment by afernandezody** [comment]
P.S. It's actually slightly more efficient setting OMP_NUM_THREADS to 2 rather than 1 but both solutions work as desired. Closing the ticket. Thanks.
